### PR TITLE
Implemented THREE.InstancedMesh

### PR DESCRIPTION
**Pull request by mrdoob**
First version of `THREE.InstancedMesh`:

```js
var count = 10000;

var mesh = new THREE.InstancedMesh( geometry, material, count );

var dummy = new THREE.Object3D();

for ( var i = 0; i < count; i ++ ) {

	dummy.position.set(
		Math.random() * 20 - 10,
		Math.random() * 20 - 10,
		Math.random() * 20 - 10
	);

	dummy.rotation.set(
		Math.random() * Math.PI,
		Math.random() * Math.PI,
		Math.random() * Math.PI
	);

	dummy.updateMatrix();

	mesh.setMatrixAt( i, dummy.matrix );

}

scene.add( mesh );
```

Screen Shot 2019-09-16 at 12 22 57 AM

Not super happy with the API yet... But it works 🙂

It's currently uploading `4 * 4 + 3 * 3 = 25` floats per instance.

If we uploaded position, quaternion and scale it would be `3 + 4 + 3 = 10` floats per instance but we'll have to run [this](https://github.com/mrdoob/three.js/blob/dev/examples/webgl_buffergeometry_instancing2.html#L38-L44) per vertex. 

Used #16141 and #16161 as reference, thanks @takahirox!

### CODE REVIEW

**Comment by looeee** [comment]
Nice!

The API look OK to me. Would these additions be possible?

``` js
var mesh = new THREE.InstancedMesh( geometry, material, count );

mesh.setPositionAt( i, x, y, z );
mesh.setScaleAt( i, x, y, z );
mesh.setRotationAt( i, x, y, z );
```

Or perhaps:

``` js
var v = new Vector3();
mesh.getPositionAt( i, v );

v.multiplyScalar( 2 );

mesh.setPositionAt( i, v );
```

**Comment by feiss** [comment]
This is great. 👍 
Will the index of each instance be available in the vertex shader? I don't see where..

**Comment by spite** [comment]
Awesome! 

A few questions:
- How are materials handled? Is every instance drawn with the same material properties?
- Can the count be modified? I imagine one can use .drawRange, but what about growing it beyond count, in case you need to dynamically shrink or expand the number of instances.
- Can the attributes be modified independently? i.e. upload only positions

For reference, this is the hacky solution i made for looper https://github.com/spite/looper/blob/master/modules/instanced.js
like this https://spite.github.io/looper/#199

**Comment by Mugen87** [comment]
More context: #5171

**Comment by mrdoob** [comment]
@feiss 

> Will the index of each instance be available in the vertex shader? I don't see where..

I don't think the instancing API support that, but we could pass a integer attribute to the shader by default? What would you like to do with it?

**Comment by mrdoob** [comment]
@spite 

> How are materials handled? Is every instance drawn with the same material properties?

Yeah. Supporting all the material properties sounds hard, but maybe color per instance to start with?

> Can the count be modified? I imagine one can use .drawRange, but what about growing it beyond count, in case you need to dynamically shrink or expand the number of instances.

I think abstracting shrinking should be easy. Expanding... not so much.

> Can the attributes be modified independently? i.e. upload only positions

I want to do performance tests comparing passing matrices to the shader vs passing position/quaternion/scale?

Part of the reason we don't have a instancing abstraction yet is because every time we try we go down the rabbit hole of "using matrices is wasteful when I only want to modify the position".

**Comment by mrdoob** [comment]
@EliasHasle 

> Will the instance matrices not respect parent transforms?

It does already.

**Comment by takahirox** [comment]
Nice work! And I'm very glad to know that somehow I helped to get `InstancedMesh` into core.

I know it's very advanced one but are you interested in `InstancedSkinnedMesh`? Each instance can differently animate with 2d texture array.

[Video](https://twitter.com/superhoge/status/1112314633543737344) / [Demo](https://raw.githack.com/takahirox/three.js/SkeletalInstancing/examples/webgl_animation_multiple_instancing_shadow.html)

If you do, I'll share my idea when I have time. (Sorry I'm busy now!)

**Comment by mrdoob** [comment]
@takahirox `InstancedSkinnedMesh` 😁?

**Comment by takahirox** [comment]
Yes. Instancing even for `SkinnedMesh`. As the video and demo show, the humanoid models are renderered at a time with instancing but they differently animate, some of them run, some others of them walk, others don't move with different speed.

image

**Comment by mrdoob** [comment]
@takahirox Okay, lets leave that idea/discussion for a different PR, but let met know if the current design doesn't work well with what `InstancedSkinnedMesh` would need.

**Comment by takahirox** [comment]
@mrdoob Yes, I wasn't going to add SkinnedMesh support into this PR but I just wanted to share my experience for future improvement.

>>  Will the index of each instance be available in the vertex shader? I don't see where..
>
>I don't think the instancing API support that, but we could pass a integer attribute to the shader by default? What would you like to do with it?

IIRC, in WebGL2 `gl_InstanceID` is assigned for each instance but in WebGL 1 it isn't. So probably we need to manually pass integers as Ricardo mentioned in WebGL 1.

>>  How are materials handled? Is every instance drawn with the same material properties?
>
> Yeah. Supporting all the material properties sounds hard, but maybe color per instance to start with?

That sounds cool!

**Comment by mrdoob** [comment]
> IIRC, in WebGL2 `gl_InstanceID` is assigned for each instance but in WebGL 1 it isn't. So probably we need to manually pass integers as Ricardo mentioned in WebGL 1.

Or maybe suggest people to use WebGL 2 for that use case 😛

**Comment by Usnul** [comment]
> If we uploaded position, quaternion and scale it would be 3 + 4 + 3 = 10 floats per instance but we'll have to run this per vertex.

From my perspective per-vertex transform is preferable, personally. It's just a bunch of multiplications and additions, fast stuff. If instances have dynamic transforms - it takes the load of computing the matrix off the CPU, and added GPU cost is very low. In addition there's less memory load, so you might end up with similar performance due to lower memory footprint (i.e. better caching).

**Comment by Usnul** [comment]
Separate concern. I would prefer if there was a way to re-size number of instances. My current implementation allows that using fairly standard over-allocation and re-allocation approach. 
When you need to grow - re-allocate memory to a larger chunk, when you shrink - just keep some unused elements within some reason until there are too many unused elements, then re-allocate. My current implementation is quite stable in this way with very few re-allocations happening as scene changes.

**Comment by takahirox** [comment]
Some thoughts for instance matrix API. In #16161 I added `.instances (Array<Object3D>)` to `InstancedMesh`. Internally Three.js calculate matrix from `instances[x]`'s position/quaternion(rotation)/scale and set to `.instancedMatrix` buffer attribute.

Pros is easy for users to control instance transform with familiar `Object3D` API, like `mesh.instances[0].rotation.x += 0.01`. Probably useful for Instancing + Animation like [this demo](https://raw.githack.com/takahirox/three.js/InstancingStandard/examples/#webgl_buffergeometry_instancing_standard) I made in #17476 

Cons is memory consumption overhead due to `Object3D` instances and performance overhead due to calculating matrices and copying to buffer attribute.

Even if this type of "easy to use API" won't get in core to avoid overhead, we may add helper in example.

**Comment by Usnul** [comment]
I think instancing is an advanced performance API, as such, I see 2 main issues to keep in mind:

1. Low per-instance memory overhead. You wouldn't use instancing if you have only a handful of objects
2. Low per-instance CPU overhead

If you are using instances - API doesn't have to be "comfortable", I believe. It should be as comfortable as it makes sense to do it, but not at the cost of performance attributes. These are my thoughts.

**Comment by takahirox** [comment]
BTW I want to make clear one thing. `Material` instance can't be shared between `InstancedMesh` and regular `Mesh` like it can't be between `SkinnedMesh` and regular `Mesh` because the shader will be specific to `InstancedMesh`, correct?

@Usnul Yeah I can understand and I know slow instancing doesn't make sense. So I lastly suggested adding helper under example not doing in core. I suppose it may be useful for some users who aren't familiar with instancing or 3D, probably helper + instancing is still faster than without instancing.

**Comment by Mugen87** [comment]
Maybe it would be good to add a second demo that demonstrates raycasting with `THREE.InstancedMesh`. It seems to me that apart from rendering instances, the interaction with single meshes is the most common task users want to do.

**Comment by pailhead** [comment]
Super long time ago there was this PR suggested, it's been working in the wild more or less like this since then, might be worth comparing notes:

https://github.com/mrdoob/three.js/pull/10750

**Comment by feiss** [comment]
> > Will the index of each instance be available in the vertex shader? I don't see where..
> 
> I don't think the instancing API support that, but we could pass a integer attribute to the shader by default? What would you like to do with it?

an integer would work. I'm thinking about setting different properties, like different colors or add vertex displacement depending on the instance. You can set different properties using the position of the vertices, but that may be not enough, or doesn't work for some effect. Does it make sense?

**Comment by 1147079942** [comment]
can this better than babylon instance.after use more than 200 instance, babylon become slow, even    
the each obj is small and each obj has about 500 count

**Comment by takahirox** [comment]
I'm not familiar with Babylon.js, but if that's true there is a chance that the reason why Babylon instancing is slow may be their comfortable API.

https://doc.babylonjs.com/how_to/how_to_use_instances

I knew their API is easy to use so suggested similar easy to use API, but I'm coming to think again that at least such API in core wouldn't be good.

**Comment by mrdoob** [comment]
I've moved the `instanceNormalMatrix` calculation to the vertex shader. Performance seems similar (faster actually) on a MacBook Pro and on a Pixel 3a.

Now that the Matrix3 is gone, it is uploading `4 * 4 = 16` floats per instance.

**Comment by mrdoob** [comment]
The example now has more monkey too 😁 

Screen Shot 2019-09-19 at 4 27 40 PM

File: examples/webgl_instancing.html [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+				var loader = new THREE.BufferGeometryLoader();
+				loader.load( 'models/json/suzanne_buffergeometry.json', function ( geometry ) {
+
+					geometry.computeVertexNormals();
+					geometry.scale( 0.5, 0.5, 0.5 );
+
+					mesh.geometry = geometry;
+
+				} );
+
```
Comment (WestLangley) [inline examples/webgl_instancing.html]: You should have only one geometry pushed to the GPU.

This works.

```js
var material = new THREE.MeshNormalMaterial();

new THREE.BufferGeometryLoader().load( 'models/json/suzanne_buffergeometry.json', function ( geometry ) {

	geometry.computeVertexNormals();
	geometry.scale( 0.5, 0.5, 0.5 );

	var mesh = new THREE.InstancedMesh( geometry, material, count );
	scene.add( mesh );

} );
```

Comment (mrdoob) [reply examples/webgl_instancing.html]: The example is still work in progress but... it actually shows that you can easily change geometries at any point, which is something that other PRs didn't make it easy.

**Comment by mrdoob** [comment]
@Mugen87 

> Maybe it would be good to add a second demo that demonstrates raycasting with `THREE.InstancedMesh`.

Implementing cpu raycasting is probably not too difficult but I wonder how fast it'll be.

File: examples/webgl_instancing.html (hunk 1 above)
Comment (WestLangley) [reply examples/webgl_instancing.html]: Then, I guess you should dispose of the old geometry...

Comment (WestLangley) [reply examples/webgl_instancing.html]: Actually, users copy these examples, so I still would suggest using the conventional pattern I posted above, and then add a GUI that allows the user to switch the geometry and/or material (if you feel it is important to demonstrate that capability). JMO.

File: src/renderers/webgl/WebGLPrograms.js [hunk 2]
```diff
@@ -134,10 +135,14 @@ function WebGLPrograms( renderer, extensions, capabilities ) {
...
 		var parameters = {
 
+			isWebGL2: capabilities.isWebGL2,
+
 			shaderID: shaderID,
 
 			precision: precision,
-			isWebGL2: capabilities.isWebGL2,
+
+			instancing: object.isInstancedMesh === true,
```
Comment (donmccurdy) [inline src/renderers/webgl/WebGLPrograms.js]: Are there other cases where reusing a single material on multiple objects would create multiple shader programs? Wondering if this should require a `material.instancing` flag or something, similar to `.skinning`.

Comment (mrdoob) [reply src/renderers/webgl/WebGLPrograms.js]: Oh, I have not tested this.

Comment (mrdoob) [reply src/renderers/webgl/WebGLPrograms.js]: I'll leave that for the next cycle.

Comment (takahirox) [reply src/renderers/webgl/WebGLPrograms.js]: I've pointed out the similar stuff [here](https://github.com/mrdoob/three.js/pull/17505#issuecomment-532012299), probably material can't be shared between regular mesh and instanced mesh.

And wondering if material can have multiple shader programs depending on some conditions, like `SkinnedMesh`, `InstancedMesh`, or so on. We can remove `.skinning` or such other properties from material and material can be shared between them.

**Comment by donmccurdy** [comment]
> > How are materials handled? Is every instance drawn with the same material properties?
> 
> Yeah. Supporting all the material properties sounds hard, but maybe color per instance to start with?

NodeMaterial actually provides a pretty decent instancing API. Modifying the transform of each instance is not as intuitive as the InstancedMesh API you've created here, but there's much more flexibility in terms of modifying any and all material properties:

```javascript
const geometry = // ... create BufferGeometry using InstancedBufferAttribute ...

const material = new StandardNodeMaterial();

const instanceID = new AttributeNode( 'instanceID', 'float' );
const instanceCount = = new FloatNode( 100 );

// Set instance positions based on instance ID.
const rootPosition = new PositionNode();
material.position = new JoinNode(
  new SwitchNode( rootPosition, 'xy' ),
  new OperatorNode( new SwitchNode( rootPosition, 'z' ), instanceID, OperatorNode.MUL ) 
);

// Set per-instance material properties from attributes.
material.color = new AttributeNode( 'instanceColor', 'vec3' );
material.roughness = new AttributeNode( 'instanceRoughness', 'float' );
```

Example: https://three-shadenodeloader.donmccurdy.com

Not sure exactly what that means for this PR, except that it would be nice if the two techniques were compatible. If instance ID and instance count from InstancedMesh were available to NodeMaterial in some reliable way, that would be a good start. 🙂

**Comment by mrdoob** [comment]
I assume it uses `InstancedBufferGeometry`? I have not changed that code so it should still work.

https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L866-L878

**Comment by donmccurdy** [comment]
Yep! I think it would still work, except that `PositionNode` probably doesn't know anything about the new instance position data yet.

File: src/objects/InstancedMesh.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+	constructor: InstancedMesh,
+
+	isInstancedMesh: true,
+
+	raycast: function () {},
+
+	setMatrixAt: function ( index, matrix ) {
+
+		matrix.toArray( this.instanceMatrix.array, index * 16 );
```
Comment (donmccurdy) [inline src/objects/InstancedMesh.js]: @mrdoob I think you meant `fromArray` here?

Comment (donmccurdy) [reply src/objects/InstancedMesh.js]: Oops no, my mistake.  😅

**Comment by takahirox** [comment]
Out of curiosity. No `InstancedMesh` document yet. Does it mean it's still an experimental and for example its API can be changed? Or just lacking of the document?

**Comment by Mugen87** [comment]
Let's add the documentation and the TS declaration file in `R110` 👍

**Comment by takahirox** [comment]
Opened an issue as a reminder #17621

**Comment by hjlld** [comment]
the current frustum culling strategy will cut off the whole `InstancedMesh`.

**Comment by mrdoob** [comment]
@hjlld yes, that's known issue. `InstancedMesh` probably needs its own `boundingSphere`...

**Comment by EliasHasle** [comment]
But `Mesh` doesn't even have a bounding sphere, does it? `boundingSphere` and `boundingBox` are on geometry. I have been thinking that sometimes it should be possible to override the default behavior per object, as an alternative to disabling frustum culling for objects that are positioned in shaders.

**Comment by Usnul** [comment]
@mrdoob you might find it interesting to have a look at how I do culling for instanced meshed in meep
https://github.com/Usnul/meep/blob/a551bba1fe41748f895af9443f820c346ea49fd4/src/model/level/foliage/InstancedFoliage.js

I don't think that you can just copy my solution, since it has a fairly large foundation in spatial indices, object pools and other stuff. But it might give you some good ideas.

PS: Word "Foliage" in the name is purely historical, that's what I was creating the system for, it is entirely just an `InstancedMesh` clone in terms of function.

**Comment by hjlld** [comment]
I resolve this by following:

Because bounding info is totally on CPU side in my project, so i split them from geometries to an indivitual map.  Then i set all `InstancedMesh` to `frustumCulled = false`, i also use BVH to make the culling strategy, so i put all the bounding infos to BVH. When BVH tells me the instance for this boudning info should be culled, i reconstruct the `InstancedMesh`, delete the corresponding matrix from `InstancedMesh.instanceMatrix` and make `InstancedMesh.count --`; in next frame, if BVH tells me the instance should be drawn, then i add the matrix and make count ++;

**Comment by EliasHasle** [comment]
@hjlld That sounds demanding for something to be done every frame.

**Comment by hookex** [comment]
When will r110  be released?  very much look forward to, urgently need to use.

**Comment by Mugen87** [comment]
> When will r110 be released?

October 30. But just to be clear, `THREE.InstancedMesh` can already be used with the current release `R109`: https://threejs.org/examples/webgl_instancing_suzanne

**Comment by hookex** [comment]
> > When will r110 be released?
> 
> October 30. But just to be clear, `THREE.InstancedMesh` can already be used with the current release `R109`: https://threejs.org/examples/webgl_instancing_suzanne

Wonderful job! Great guys!

File: examples/webgl_instancing.html [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+								mesh.setMatrixAt( i ++, dummy.matrix );
+
+							}
+
+						}
+
+					}
+
+					mesh.instanceMatrix.needsUpdate = true;
```
Comment (Mugen87) [inline examples/webgl_instancing.html]: FYI: I've used `THREE.InstancedMesh` in a new project and realized it makes sense to add the following line when creating the mesh if `instanceMatrix` is updated per frame like in this example.
```js
mesh.instanceMatrix.setUsage( THREE.DynamicDrawUsage );
```

**Comment by nivashmani91** [comment]
Hi @mrdoob, Is there any possible ways to implement the instancing for lines and sprites?.

**Comment by Mugen87** [comment]
@nivashmani91 Please read #25078 for more information about that topic.

**Pull request merged**